### PR TITLE
DEV: Use runtime info to split test files for parallel testing

### PR DESCRIPTION
**Pull request by tgxworld**
Using the runtime information, we will be able to more efficiently group
the test files across the test processes hence leading to better
utilization of resources.

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -173,11 +173,11 @@ jobs:
 
       - name: Core RSpec
         if: matrix.build_type == 'backend' && matrix.target == 'core'
-        run: bin/turbo_rspec --verbose
+        run: bin/turbo_rspec --use-runtime-info --verbose
```
Comment (tgxworld) [inline .github/workflows/tests.yml:176]: Note that we no files are passed as arguments, runtime info is enabled by default but I'm adding the `--use-runtime-info` flag for clarity sake.

**Comment by tgxworld** [comment]
Note that this will be something we have to test out since the benefits only comes from subsequent runs.

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -150,10 +150,11 @@ jobs:
       - name: Fetch turbo_rspec_runtime.log cache
         uses: actions/cache@v3
         id: test-runtime-cache
-        if: matrix.build_type == 'backend' && matrix.target == 'core'
+        if: matrix.build_type == 'backend' || matrix.build_type == 'system'
         with:
           path: tmp/turbo_rspec_runtime.log
-          key: rspec-runtime-backend-core
+          key: rspec-runtime-${{ matrix.build_type }}-${{ matrix.target }}-${{ github.run_id }}
+          restore-keys: rspec-runtime-${{ matrix.build_type }}-${{ matrix.target }}
```
Comment (tgxworld) [inline .github/workflows/tests.yml:157]: Accordingly to https://github.com/actions/cache/blob/04f198bf0b2a39f7230a4304bf07747a0bddf146/tips-and-workarounds.md#update-a-cache, this allows us to update the cache every run and restore from the latest.

**Review by gschlager — APPROVED** [review]

**Comment by gschlager** [comment]
> Just wondering where the log is stored and read from between CI runs, if it's just going to a `tmp` folder? Do we need to persist + read it somehow or is it automatic?

The caching workflow step takes care of it.

https://github.com/discourse/discourse/blob/cc87f3d1a928b92db14250c0cba0c72f0798de43/.github/workflows/tests.yml#L150-L157

For example, take a look at https://github.com/discourse/discourse/actions/runs/5150488818/jobs/9274680993. You'll see "Fetch turbo_rspec_runtime.log cache" and "Post Fetch turbo_rspec_runtime.log cache" in the output.

**Comment by martin-brennan** [comment]
Nice thanks for that @gschlager !

**Comment by tgxworld** [comment]
Thank you for the review @gschlager :+1:

**Pull request merged**